File: jdroid-java-core/src/main/java/com/jdroid/java/repository/InMemoryRepository.java
package com.jdroid.java.repository;

import com.jdroid.java.collections.Lists;
import com.jdroid.java.collections.Maps;
import com.jdroid.java.domain.Identifiable;
import com.jdroid.java.exception.UnexpectedException;
import com.jdroid.java.utils.LoggerUtils;
import com.jdroid.java.utils.ReflectionUtils;

import org.slf4j.Logger;

import java.util.Collection;
import java.util.List;
import java.util.Map;

/**
 * @param <T>
 */
public class InMemoryRepository<T extends Identifiable> implements Repository<T> {

	private static final Logger LOGGER = LoggerUtils.getLogger(InMemoryRepository.class);

	private long nextId = 1;
	private Map<String, T> items = Maps.INSTANCE.newLinkedHashMap();

	@Override
	public void add(T item) {
		if (item.getId() == null) {
			ReflectionUtils.setId(item, nextId++);
		}
		items.put(item.getId(), item);
		LOGGER.debug("Added object in memory: " + item);
	}

	@Override
	public void addAll(Collection<T> items) {
		for (T item : items) {
			add(item);
		}
		LOGGER.debug("Stored objects in memory:\n" + items);
	}

	@Override
	public void update(T item) {
		if (item.getId() == null) {
			throw new UnexpectedException("Item with null id can not be updated");
		}
		items.put(item.getId(), item);
		LOGGER.debug("Updated object in memory: " + item);
	}

	@Override
	public void remove(T item) {
		items.remove(item.getId());
		LOGGER.debug("Deleted object from memory: " + item);
	}

	@Override
	public void removeAll(Collection<T> items) {
		for (T item : items) {
			remove(item);
		}
		LOGGER.debug("Deleted objects in memory: " + items);
	}

	@Override
	public void replaceAll(Collection<T> items) {
		removeAll();
		addAll(items);
	}

	@Override
	public List<T> getAll() {
		List<T> results = Lists.INSTANCE.newArrayList(items.values());
		LOGGER.debug("Retrieved all objects [" + results.size() + "] from memory");
		return results;
	}

	@Override
	public T get(String id) {
		return items.get(id);
	}

	@Override
	public void removeAll() {
		items.clear();
		LOGGER.debug("Deleted from memory all objects from this repository");
	}

	@Override
	public void remove(String id) {
		items.remove(id);
	}

	@Override
	public Boolean isEmpty() {
		return items.isEmpty();
	}

	@Override
	public Long getSize() {
		return (long)items.size();
	}

	@Override
	public List<T> getByField(String fieldName, Object... values) {
		throw new UnsupportedOperationException();
	}

	@Override
	public T getItemByField(String fieldName, Object... values) {
		throw new UnsupportedOperationException();
	}

	@Override
	public List<T> getByIds(List<String> ids) {
		List<T> itemsList = Lists.INSTANCE.newArrayList();
		for (String each : ids) {
			T item = items.get(each);
			if (item != null) {
				itemsList.add(item);
			}
		}
		return itemsList;
	}

	@Override
	public T getUniqueInstance() {
		return items.isEmpty() ? null : items.values().iterator().next();
	}
}

File: jdroid-java-core/src/main/java/com/jdroid/java/repository/InMemoryRepository.kt
package com.jdroid.java.repository

import com.jdroid.java.collections.Lists
import com.jdroid.java.collections.Maps
import com.jdroid.java.domain.Identifiable
import com.jdroid.java.exception.UnexpectedException
import com.jdroid.java.utils.LoggerUtils
import com.jdroid.java.utils.ReflectionUtils

open class InMemoryRepository<T : Identifiable> : Repository<T> {

    companion object {
        private val LOGGER = LoggerUtils.getLogger(InMemoryRepository::class.java)
    }

    private var nextId: Long = 1
    private val items = Maps.newLinkedHashMap<String, T>()

    override fun add(item: T) {
        if (item.getId() == null) {
            ReflectionUtils.setId(item, nextId++)
        }
        items[item.getId()!!] = item
        LOGGER.debug("Added object in memory: $item")
    }

    override fun addAll(items: Collection<T>) {
        for (item in items) {
            add(item)
        }
        LOGGER.debug("Stored objects in memory:\n$items")
    }

    override fun update(item: T) {
        if (item.getId() == null) {
            throw UnexpectedException("Item with null id can not be updated")
        }
        items[item.getId()!!] = item
        LOGGER.debug("Updated object in memory: $item")
    }

    override fun remove(item: T) {
        items.remove(item.getId())
        LOGGER.debug("Deleted object from memory: $item")
    }

    override fun removeAll(items: Collection<T>) {
        for (item in items) {
            remove(item)
        }
        LOGGER.debug("Deleted objects in memory: $items")
    }

    override fun replaceAll(items: Collection<T>) {
        removeAll()
        addAll(items)
    }

    override fun getAll(): List<T> {
        val results = Lists.newArrayList(items.values)
        LOGGER.debug("Retrieved all objects [" + results.size + "] from memory")
        return results
    }

    override fun get(id: String): T? {
        return items[id]
    }

    override fun removeAll() {
        items.clear()
        LOGGER.debug("Deleted from memory all objects from this repository")
    }

    override fun remove(id: String) {
        items.remove(id)
    }

    override fun isEmpty(): Boolean {
        return items.isEmpty()
    }

    override fun getSize(): Long {
        return items.size.toLong()
    }

    override fun getByField(fieldName: String, vararg values: Any): List<T> {
        throw UnsupportedOperationException()
    }

    override fun getItemByField(fieldName: String, vararg values: Any): T? {
        throw UnsupportedOperationException()
    }

    override fun getByIds(ids: List<String>): List<T> {
        val itemsList = mutableListOf<T>()
        for (each in ids) {
            val item = items[each]
            if (item != null) {
                itemsList.add(item)
            }
        }
        return itemsList
    }

    override fun getUniqueInstance(): T? {
        return if (items.isEmpty()) null else items.values.iterator().next()
    }
}
